Fix ActionMailer assertion not working for mail defining delivery_job:

Edouard-chin opened the pull request:
Fix ActionMailer assertion not working for mail defining delivery_job:

- If a Mail defines a custom delivery_job, all ActionMailer assertion
      helper (assert_emails, assert_enqueued_emails ...) wouldn't work.

 ```ruby
MyMailer < ApplicationMailer
  self.delivery_job = MyJob
end

# This assertion will fail
assert_emails(1) do
  MyMailer.my_mail.deliver_later
end
```

This PR leverage the new ActiveJob feature that accepts Procs for the
      `only` keyword and check if the delivery job is one of ActionMailer
       registered ones.

kamipo commented on actionmailer/lib/action_mailer/test_adapter.rb:
```diff
@@ -0,0 +1,23 @@
+  class TestAdapter < ActiveJob::QueueAdapters::TestAdapter
+    private
+
+      def filtered?(job) # :nodoc:
```
`:nodoc:` for private methods is redundant.

Edouard-chin commented on actionmailer/lib/action_mailer/test_helper.rb:
```diff
@@ -151,7 +165,13 @@ def assert_enqueued_email_with(mailer, method, args: nil, queue: "mailers", &blo
+    private
+
+      def queue_adapter_for_test
+        ActionMailer::TestAdapter.new
```
The mailer_adapter has the precedence, active_job assertions would just behave as usual though.

gmcgibbon replied:
Right, I can see regular job assertions run through `ActionMailer::TestAdapter` now. I'm not sure I'm a huge fan of that. Do you think it makes sense to change this behaviour?

Edouard-chin replied:
Having 2 adapters is even more confusing IMO. If we want to avoid having an extra adapter to fix this issue I think we don't have the choice but to implement lambda filter. @rafaelfranca got some opinion/advice on this one ?

rafaelfranca replied:
Allowing the filter to be a lambda and building the lambda in `assert_enqueued_emails` and friends makes more sense to me

Edouard-chin commented on actionmailer/lib/action_mailer/test_helper.rb:
```diff
@@ -1,13 +1,16 @@
   module TestHelper
     include ActiveJob::TestHelper
 
+    DELIVERY_JOB = Object.new
```
I thought about the lambda filter and had doubts so I went for an implementation that don't reach into the ActiveJob boundaries. My doubt was about the usefulness of supporting lambdas by applications.
If we decide to add this option to help fix the ActionMailer issue this will become public API, and again I wasn't sure if that's something developers will benefit from. Thoughts?

Edouard-chin committed:
Allow all ActiveJob assertion helper to accept Proc in their `only` kw:

- That feature is useful to enqueue or assert that jobs got enqueued
  or performed based on dynamic conditions.
  We will be able to leverage that feature to fix all ActionMailer
  assertion helper issue when a Mailer define a custom delivery job
  (see next commit).

Edouard-chin committed:
Fix ActionMailer assertion not working for mail defining delivery_job:

- If a Mail defines a custom delivery_job, all ActionMailer assertion
  helper (assert_emails, assert_enqueued_emails ...) wouldn't work.

  ```ruby
    MyMailer < ApplicationMailer
      self.delivery_job = MyJob
    end

    # This assertion will fail
    assert_emails(1) do
      MyMailer.my_mail.deliver_later
    end

  This PR leverage the new ActiveJob feature that accepts Procs for the
  `only` keyword and check if the delivery job is one of ActionMailer
   registered ones.

Edouard-chin commented:
Sorry for the delay, this is ready with a new implementation
